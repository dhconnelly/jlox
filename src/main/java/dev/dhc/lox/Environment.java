package dev.dhc.lox;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class Environment {
  private final Optional<Environment> up;
  private final Map<String, Value> values = new HashMap<>();

  public Environment() {
    up = Optional.empty();
  }

  public Environment(Environment up) {
    this.up = Optional.of(up);
  }

  public void define(String name, Value value) {
    values.put(name, value);
  }

  public Optional<Value> assign(String name, Value value) {
    if (values.containsKey(name)) {
      values.put(name, value);
      return Optional.of(value);
    }
    return up.flatMap(env -> env.assign(name, value));
  }

  public Optional<Value> get(String name) {
    final var value = values.get(name);
    if (value != null) return Optional.of(value);
    if (up.isPresent()) return up.get().get(name);
    return Optional.empty();
  }

  private Environment up(int depth) {
    var env = this;
    for (int i = 0; i < depth; i++) {
      env = env.up.get();
    }
    return env;
  }

  public Value assignAt(int depth, String name, Value value) {
    up(depth).values.put(name, value);
    return value;
  }

  public Value getAt(int depth, String name) {
    return up(depth).values.get(name);
  }
}
