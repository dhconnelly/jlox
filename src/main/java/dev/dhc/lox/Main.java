package dev.dhc.lox;

import static java.nio.charset.StandardCharsets.UTF_8;

import dev.dhc.lox.Token.Type;
import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.nio.file.Files;
import java.nio.file.Paths;

public class Main {
  public record Environment(InputStream in, PrintStream out, PrintStream err) {}

  public static int run(Environment env, String[] args) throws IOException {
    if (args.length == 0) {
      env.err.println("usage: lox COMMAND [ARG...]");
      return 64;
    }

    final var command = args[0];
    switch (command) {
      case "tokenize" -> {
        if (args.length != 2) {
          env.err.println("usage: lox tokenize FILE");
          return 64;
        }
        final var in = Files.newInputStream(Paths.get(args[1]));
        final var scanner = new Scanner(in);
        boolean hadError = false;
        while (true) {
          try {
            final var token = scanner.nextToken();
            env.out.println(token);
            if (token.type() == Type.EOF) break;
          } catch (LoxError e) {
            env.err.println(e.getMessage());
            hadError = true;
          }
        }
        return hadError ? 65 : 0;
      }

      case "repl" -> {
        final var reader = new BufferedReader(new InputStreamReader(env.in));
        boolean hadError = false;
        while (true) {
          env.out.print("> ");
          final var line = reader.readLine();
          if (line == null) break;
          final var scanner = new Scanner(new ByteArrayInputStream(line.getBytes(UTF_8)));
          try {
            while (true) {
              final var token = scanner.nextToken();
              env.out.println(token);
              if (token.type() == Type.EOF) break;
            }
          } catch (LoxError e) {
            env.err.println(e.getMessage());
            hadError = true;
          }
        }
        return hadError ? 65 : 0;
      }

      case "run" -> {
        if (args.length != 2) {
          env.err.println("usage: lox run FILE");
          return 64;
        }
        final var in = Files.newInputStream(Paths.get(args[1]));
        final var scanner = new Scanner(in);
        try {
          while (true) {
            final var token = scanner.nextToken();
            env.out.println(token);
            if (token.type() == Type.EOF) break;
          }
        } catch (LoxError e) {
          env.err.println(e.getMessage());
          return 65;
        }
        return 0;
      }

      default -> {
        env.err.println("unrecognized command: " + command);
        return 64;
      }
    }
  }

  public static void main(String[] args) throws IOException {
    final var env = new Environment(System.in, System.out, System.err);
    System.exit(run(env, args));
  }
}
