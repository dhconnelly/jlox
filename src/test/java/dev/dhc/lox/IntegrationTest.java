package dev.dhc.lox;

import static java.nio.charset.StandardCharsets.UTF_8;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import dev.dhc.lox.Main.Environment;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.PrintStream;
import java.nio.file.Paths;
import java.util.List;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.ValueSource;

@ExtendWith({SnapshotExtension.class})
public class IntegrationTest {
  private record Result(int code, List<String> outLines, List<String> errLines) {}

  private Result execute(String... args) throws IOException {
    final var in = new ByteArrayInputStream(new byte[]{});
    final var out = new ByteArrayOutputStream();
    final var err = new ByteArrayOutputStream();
    final var exit = Main.run(
        new Environment(in, new PrintStream(out), new PrintStream(err)), args);
    return new Result(
        exit,
        out.toString(UTF_8).lines().toList(),
        err.toString(UTF_8).lines().toList());
  }

  private String resourcePath(String resource) {
    final var url = getClass().getClassLoader().getResource(resource);
    if (url == null) throw new RuntimeException("resource not found: " + resource);
    return Paths.get(url.getPath()).toFile().getAbsolutePath();
  }

  private Expect expect;

  @ParameterizedTest
  @ValueSource(strings = {
      "inputs/scanner/test.lox",
      "inputs/scanner/empty.lox",
      "inputs/scanner/tokens.lox",
      "inputs/scanner/scanner_errors.lox"
  })
  void testScanner(String resource) throws IOException {
    expect.scenario(resource).toMatchSnapshot(execute("tokenize", resourcePath(resource)));
  }

  @ParameterizedTest
  @ValueSource(strings = {
      "inputs/parser/test.lox",
      "inputs/parser/empty.lox",
      "inputs/parser/expressions.lox",
      "inputs/parser/parser_errors.lox",
  })
  void testParser(String resource) throws IOException {
    expect.scenario(resource).toMatchSnapshot(execute("parse", resourcePath(resource)));
  }

  @ParameterizedTest
  @ValueSource(strings = {
      "inputs/evaluator/empty.lox",
      "inputs/evaluator/values.lox",
  })
  void testEvaluator(String resource) throws IOException {
    expect.scenario(resource).toMatchSnapshot(execute("evaluate", resourcePath(resource)));
  }
}
