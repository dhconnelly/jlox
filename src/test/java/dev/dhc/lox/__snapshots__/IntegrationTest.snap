dev.dhc.lox.IntegrationTest.testScanner[empty.lox]=[
  {
    "code": 0,
    "outLines": [
      "EOF  null"
    ],
    "errLines": [ ]
  }
]


dev.dhc.lox.IntegrationTest.testScanner[scanner_errors.lox]=[
  {
    "code": 65,
    "outLines": [
      "EOF  null"
    ],
    "errLines": [
      "[line 1] Error: Unexpected character: #",
      "[line 2] Error: Unterminated string."
    ]
  }
]


dev.dhc.lox.IntegrationTest.testScanner[test.lox]=[
  {
    "code": 0,
    "outLines": [
      "VAR var null",
      "IDENTIFIER language null",
      "EQUAL = null",
      "STRING \"lox\" lox",
      "SEMICOLON ; null",
      "EOF  null"
    ],
    "errLines": [ ]
  }
]


dev.dhc.lox.IntegrationTest.testScanner[tokens.lox]=[
  {
    "code": 0,
    "outLines": [
      "LEFT_PAREN ( null",
      "LEFT_PAREN ( null",
      "RIGHT_PAREN ) null",
      "LEFT_BRACE { null",
      "LEFT_BRACE { null",
      "LEFT_BRACE { null",
      "RIGHT_BRACE } null",
      "RIGHT_BRACE } null",
      "LEFT_BRACE { null",
      "RIGHT_BRACE } null",
      "LEFT_BRACE { null",
      "LEFT_BRACE { null",
      "LEFT_BRACE { null",
      "RIGHT_BRACE } null",
      "COMMA , null",
      "DOT . null",
      "MINUS - null",
      "PLUS + null",
      "SEMICOLON ; null",
      "STAR * null",
      "EQUAL = null",
      "EQUAL_EQUAL == null",
      "EQUAL_EQUAL == null",
      "EQUAL = null",
      "EQUAL_EQUAL == null",
      "EQUAL_EQUAL == null",
      "EQUAL_EQUAL == null",
      "EQUAL_EQUAL == null",
      "EQUAL = null",
      "BANG_EQUAL != null",
      "BANG ! null",
      "LESS < null",
      "GREATER > null",
      "LESS_EQUAL <= null",
      "GREATER_EQUAL >= null",
      "EQUAL = null",
      "SLASH / null",
      "SLASH / null",
      "STRING \"\" ",
      "STRING \"foo\" foo",
      "STRING \"foo bar baz\" foo bar baz",
      "STRING \"123.45\" 123.45",
      "NUMBER 1 1.0",
      "NUMBER 1.2 1.2",
      "NUMBER 123.45 123.45",
      "NUMBER 0 0.0",
      "NUMBER 0.0 0.0",
      "IDENTIFIER foo null",
      "IDENTIFIER bar null",
      "IDENTIFIER baz null",
      "AND and null",
      "OR or null",
      "IDENTIFIER andor null",
      "CLASS class null",
      "ELSE else null",
      "FALSE false null",
      "FUN fun null",
      "FOR for null",
      "IF if null",
      "NIL nil null",
      "PRINT print null",
      "RETURN return null",
      "SUPER super null",
      "THIS this null",
      "TRUE true null",
      "VAR var null",
      "WHILE while null",
      "EOF  null"
    ],
    "errLines": [ ]
  }
]